File: labAA/STR2.cpp
#include <iostream>
#include <unordered_set>
#include <vector>

using namespace std;

vector<int> countDistinctSubstrings(const string& str) {
    int n = str.length();
    vector<int> result(n);

    unordered_set<string> substrings;

    for (int len = 1; len <= n; len++) {
        substrings.clear();

        for (int i = 0; i <= n - len; i++) {
            string substring = str.substr(i, len);
            substrings.insert(substring);
        }

        result[len - 1] = substrings.size();
    }

    return result;
}

int main() {
    string str;
    cin >> str;

    vector<int> distinctSubstrings = countDistinctSubstrings(str);

    for (int count : distinctSubstrings) {
        cout << count << " ";
    }

    cout << endl;

    return 0;
}

File: labAA/STR2.py
def count_distinct_substrings(string):
    n = len(string)
    suffix_array = sorted(range(n), key=lambda i: string[i:])
    
    lcp_array = [0] * n
    for i in range(1, n):
        x, y = suffix_array[i-1], suffix_array[i]
        while x < n and y < n and string[x] == string[y]:
            lcp_array[i] += 1
            x += 1
            y += 1
    
    distinct_substrings = [0] * n
    for i in range(1, n):
        distinct_substrings[suffix_array[i]] = n - suffix_array[i]
    
    for i in range(1, n):
        distinct_substrings[i] -= lcp_array[i]
    
    for count in distinct_substrings[1:]:
        print(count, end=" ")
    print(0)  # Last substring of length n has 0 distinct substrings

# Example usage
string = "mississippi"
count_distinct_substrings(string)
